rotkehlchen/chain/evm/decoding/curve/decoder.py

```diff
@@ -0,0 +1,732 @@
+import logging
+from collections.abc import Callable
+from typing import TYPE_CHECKING, Any, Literal
+
+from rotkehlchen.accounting.structures.balance import Balance
+from rotkehlchen.assets.asset import Asset
+from rotkehlchen.chain.ethereum.utils import asset_normalized_value
+from rotkehlchen.chain.evm.constants import ETH_SPECIAL_ADDRESS, ZERO_ADDRESS
+from rotkehlchen.chain.evm.decoding.aave.constants import CPT_AAVE_V1, CPT_AAVE_V2, CPT_AAVE_V3
+from rotkehlchen.chain.evm.decoding.curve.constants import (
+    ADD_LIQUIDITY_EVENTS,
+    CPT_CURVE,
+    EXCHANGE_MULTIPLE,
+    GAUGE_DEPOSIT,
+    GAUGE_VOTE,
+    GAUGE_WITHDRAW,
+    REMOVE_LIQUIDITY_EVENTS,
+    REMOVE_LIQUIDITY_IMBALANCE,
+    TOKEN_EXCHANGE,
+    TOKEN_EXCHANGE_UNDERLYING,
+)
+from rotkehlchen.chain.evm.decoding.curve.curve_cache import (
+    get_query_curve_data,
+    get_read_curve_pools_and_gauges,
+    get_save_curve_data_to_cache,
+)
+from rotkehlchen.chain.evm.decoding.interfaces import (
+    DecoderInterface,
+    ReloadablePoolsAndGaugesDecoderMixin,
+)
+from rotkehlchen.chain.evm.decoding.structures import (
+    DEFAULT_DECODING_OUTPUT,
+    FAILED_ENRICHMENT_OUTPUT,
+    DecoderContext,
+    DecodingOutput,
+    EnricherContext,
+    TransferEnrichmentOutput,
+)
+from rotkehlchen.chain.evm.decoding.types import CounterpartyDetails
+from rotkehlchen.chain.evm.decoding.utils import maybe_reshuffle_events
+from rotkehlchen.chain.evm.structures import EvmTxReceiptLog
+from rotkehlchen.constants.assets import A_ETH
+from rotkehlchen.constants.resolver import evm_address_to_identifier
+from rotkehlchen.errors.asset import UnknownAsset, WrongAssetType
+from rotkehlchen.history.events.structures.evm_event import EvmProduct
+from rotkehlchen.history.events.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import CacheType, ChainID, ChecksumEvmAddress, EvmTokenKind, EvmTransaction
+from rotkehlchen.utils.misc import hex_or_bytes_to_address, hex_or_bytes_to_int
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.history.events.structures.evm_event import EvmEvent
+    from rotkehlchen.user_messages import MessagesAggregator
+
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+def _read_curve_asset(
+        asset_address: ChecksumEvmAddress | None,
+        chain_id: ChainID,
+) -> Asset | None:
+    """
+    A thin wrapper that turns asset address into an asset object.
+
+    Object returned here is a pure Asset (not a resolved CryptoAsset) since it is meant only for
+    comparison with other assets. And to compare with other assets there is no need to resolve.
+    """
+    if asset_address is None:
+        return None
+
```

Comment:
when do you receive address None? Without reading the code it looks more as an error than something yu want

Reply:
I found it coming as None in these tests:
```
test_curve_trade_token_to_token
test_curve_trade_eth_to_token
test_curve_trade_exchange_underlying
```
It seems to be a possible case because it's also explained in the docstring of `_decode_curve_trades`